torch/_inductor/codegen/wrapper.py

```diff
@@ -1958,28 +1972,96 @@ def codegen_unbacked_symbol_decl(self, symbol):
+            with V.set_graph_handler(subgraph.graph):
+                subgraph.graph.codegen_subgraph(
+                    parent_graph=parent_graph,
+                )
+            _codegen_subgraph_suffix()
+        finally:
+            self.pop_codegened_graph()
+
     def codegen_subgraph_prefix(self, subgraph, outer_inputs, outer_outputs):
+        subgraph.graph.add_symbol_graph_inputs()
         for inner_input, outer_input in zip(subgraph.graph.graph_inputs, outer_inputs):
             self.writeline(f"{self.declare}{inner_input} = {outer_input}{self.ending}")
 
     def codegen_subgraph_suffix(self, subgraph, outer_inputs, outer_outputs):
         for inner_output, outer_output in zip(
-            subgraph.graph.graph_outputs, outer_outputs
+            subgraph.graph.get_output_names(), outer_outputs
         ):
-            self.writeline(
-                f"{outer_output} = {inner_output.codegen_reference()}{self.ending}"
-            )
+            self.writeline(f"{outer_output} = {inner_output}{self.ending}")
+
+    def codegen_subgraph_call(self, subgraph, outer_inputs, outer_outputs):
+        # Get the input and output names of the subgraph
+        inner_inputs = ", ".join(subgraph.graph.graph_input_names)
+        if len(subgraph.graph.graph_input_names) == 1:
+            inner_inputs += ","
+
+        output_names = subgraph.graph.get_output_names()
+        inner_outputs = ", ".join(output_names)
+        if len(output_names) == 1:
+            inner_outputs += ","
+
+        # Call the subgraph launcher function
+        self.writeline(f"({inner_outputs}) = {subgraph.graph.name}([{inner_inputs}])")
 
     def codegen_subgraph(self, subgraph, outer_inputs, outer_outputs):
+        # Codegen subgraph by recursively calling the codegen for the subgraph.
+        # This lifts the subgraph as a function in the output code.
+        if V.graph.aot_mode:
+            self.codegen_subgraph_by_inlining(subgraph, outer_inputs, outer_outputs)
+            return
         try:
             self.push_codegened_graph(subgraph.graph)
             self.writeline(f"{self.comment} subgraph: {subgraph.name}")
             self.codegen_subgraph_prefix(subgraph, outer_inputs, outer_outputs)
+
             parent_graph = V.graph
-            with V.set_graph_handler(subgraph.graph):
-                subgraph.graph.codegen_subgraph(
-                    parent_graph=parent_graph,
-                )
+            subgraph.graph.cpp_wrapper = parent_graph.cpp_wrapper
+
+            if subgraph.graph.name not in self.already_codegened_subgraphs:
+                # If its codegened, the parent wrapper already has subgraph fn by name subgraph.graph.name
+                with V.set_graph_handler(subgraph.graph):
+                    # Call the codegen of subgraph recursively
+                    subgraph_code = subgraph.graph.codegen()[0]
+                    self.already_codegened_subgraphs.add(subgraph.graph.name)
```

Comment:
Should this line (and below) be under the context manager?

Reply:
I moved the line out of the context manager.